File: CeliTax/BuilderFactory.m
//
//  BuilderFactory.m
//  CeliTax
//
//  Created by Leon Chen on 2015-07-20.
//  Copyright (c) 2015 CraveNSave. All rights reserved.
//

#import "BuilderFactory.h"
#import "CatagoryBuilder.h"
#import "Catagory.h"
#import "RecordBuilder.h"
#import "Record.h"
#import "Receipt.h"
#import "ReceiptBuilder.h"
#import "TaxYear.h"
#import "TaxYearBuilder.h"

@interface BuilderFactory ()

@property (nonatomic, strong) CatagoryBuilder *catagoryBuilder;
@property (nonatomic, strong) RecordBuilder *recordBuilder;
@property (nonatomic, strong) ReceiptBuilder *receiptBuilder;
@property (nonatomic, strong) TaxYearBuilder *taxYearBuilder;

@end

@implementation BuilderFactory

- (CatagoryBuilder *) createCatagoryBuilder
{
    if (self.catagoryBuilder)
    {
        return self.catagoryBuilder;
    }
    
    self.catagoryBuilder = [[CatagoryBuilder alloc] init];
    
    return self.catagoryBuilder;
}

- (RecordBuilder *) createRecordBuilder
{
    if (self.recordBuilder)
    {
        return self.recordBuilder;
    }
    
    self.recordBuilder = [[RecordBuilder alloc] init];
    
    return self.recordBuilder;
}

- (ReceiptBuilder *) createReceiptBuilder
{
    if (self.receiptBuilder)
    {
        return self.receiptBuilder;
    }
    
    self.receiptBuilder = [[ReceiptBuilder alloc] init];
    
    return self.receiptBuilder;
}

- (TaxYearBuilder *) createTaxYearBuilder
{
    if (self.taxYearBuilder)
    {
        return self.taxYearBuilder;
    }
    
    self.taxYearBuilder = [[TaxYearBuilder alloc] init];
    
    return self.taxYearBuilder;
}


@end

File: CeliTax/BuilderFactory.swift
//
//  BuilderFactory.swift
//  CeliTax
//
//  Created by Leon Chen on 2015-11-01.
//  Copyright © 2015 CraveNSave. All rights reserved.
//

import Foundation

@objc
class BuilderFactory : NSObject //TODO: Remove Subclass to NSObject when the entire app has been converted to Swift
{
    var categoryBuilder : CategoryBuilder?
    var recordBuilder : RecordBuilder?
    var receiptBuilder : ReceiptBuilder?
    var taxYearBuilder : TaxYearBuilder?
    
    func createCatagoryBuilder() -> CategoryBuilder
    {
        if let categoryBuilder = self.categoryBuilder
        {
            return categoryBuilder
        }
        
        self.categoryBuilder = CategoryBuilder()
        
        return self.categoryBuilder!
    }
    
    func createRecordBuilder() -> RecordBuilder
    {
        if (self.recordBuilder != nil)
        {
            return self.recordBuilder!
        }
        
        self.recordBuilder = RecordBuilder()
        
        return self.recordBuilder!
    }
    
    func createReceiptBuilder() -> ReceiptBuilder
    {
        if (self.receiptBuilder != nil)
        {
            return self.receiptBuilder!
        }
        
        self.receiptBuilder = ReceiptBuilder()
        
        return self.receiptBuilder!
    }
    
    func createTaxYearBuilder() -> TaxYearBuilder
    {
        if (self.taxYearBuilder != nil)
        {
            return self.taxYearBuilder!
        }
        
        self.taxYearBuilder = TaxYearBuilder()
        
        return self.taxYearBuilder!
    }
}